dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java

```diff
@@ -34,12 +36,22 @@
  */
 public class StoryBlockAPIImpl implements StoryBlockAPI {
 
-    @CloseDBIfOpened
+    private static final Lazy<Integer> MAX_RECURSION_LEVEL = Lazy.of(() -> Config.getIntProperty("STORY_BLOCK_MAX_RECURSION_LEVEL", 2));
+
     @Override
+    @CloseDBIfOpened
     public StoryBlockReferenceResult refreshReferences(final Contentlet contentlet) {
         final MutableBoolean refreshed = new MutableBoolean(false);
         if (null != contentlet && null != contentlet.getContentType() &&
                 contentlet.getContentType().hasStoryBlockFields()) {
+
+            if (ExceptionUtil.isMethodCallCountGteThan(
```

Comment:
don't love this logic - it pulls 1000 lines of the thread stack and concatenates them together for a new String, which is gonna be huge, and then does a find?  

Just loop over the thread stack[] and search for the string in the line and when that number is >= the MAX_RECURSION, then break.  No new Strings allocated.

Reply:
Done!